src/compile.ts

```diff
@@ -55,38 +58,103 @@ export function compile(
+          } catch (err) {
+            const error =
+              err instanceof Error ? err : Error("Unknown compiler error.");
+            return newExpr("Err", [
+              ts.factory.createNewExpression(
+                ts.factory.createIdentifier(error.name),
+                undefined,
+                [ts.factory.createStringLiteral(error.message)]
+              ),
+            ]) as unknown as Statement;
           }
-        } catch (err) {
-          const error =
-            err instanceof Error ? err : Error("Unknown compiler error.");
-          return newExpr("Err", [
-            ts.factory.createNewExpression(
-              ts.factory.createIdentifier(error.name),
-              undefined,
-              [ts.factory.createStringLiteral(error.message)]
-            ),
-          ]) as unknown as Statement;
-        }
-        return ts.visitEachChild(node, visitor, ctx);
+        };
+        // keep processing the children of the updated node.
+        return ts.visitEachChild(_visit(), visitor, ctx);
+      }
+
+      /**
+       * Various types that could be in a call argument position of a function parameter.
+       */
+      type TsFunctionParameter =
+        | ts.FunctionExpression
+        | ts.ArrowFunction
+        | ts.Identifier
+        | ts.PropertyAccessExpression
+        | ts.ElementAccessExpression
+        | ts.CallExpression;
+
+      type EventBusWhenInterface = ts.CallExpression & {
+        arguments: [any, any, TsFunctionParameter];
+      };
+
+      type EventBusMapInterface = ts.CallExpression & {
+        arguments: [TsFunctionParameter];
+      };
+
+      function isEventBusWhenFunction(
+        node: ts.Node
+      ): node is EventBusWhenInterface {
+        return (
+          ts.isCallExpression(node) &&
+          ts.isPropertyAccessExpression(node.expression) &&
+          node.expression.name.text === "when" &&
+          isEventBus(node.expression.expression)
+        );
+      }
+
+      function isEventBusMapFunction(
+        node: ts.Node
+      ): node is EventBusMapInterface {
+        return (
+          ts.isCallExpression(node) &&
+          ts.isPropertyAccessExpression(node.expression) &&
+          node.expression.name.text === "map" &&
+          isEventBusRule(node.expression.expression)
+        );
+      }
+
+      /**
+       * Checks to see if a node is of type EventBus.
+       * The node could be any kind of node that returns an event bus rule.
+       */
+      function isEventBus(node: ts.Node) {
+        return isFunctionlessClassOfKind(node, EventBus.FunctionlessType);
+      }
+
+      /**
+       * Checks to see if a node is of type EventBusRule.
+       * The node could be any kind of node that returns an event bus rule.
+       */
+      function isEventBusRule(node: ts.Node) {
+        return isFunctionlessClassOfKind(node, EventBusRule.FunctionlessType);
+      }
+
+      function isFunctionlessClassOfKind(node: ts.Node, kind: string) {
```

Comment:
Running this multiple times per node is wasteful.

Reply:
Ehh? what part of it? Seems like early optimization honestly. We could assign some form of name to a node type, but there could be multiple layers of validation that differ between node types, including which level needs to be checked.

For example, `isEventBusMapFunction` checks it for `node.expression.expression` only when `node.expression` is named map. 

`isNewEventBusRule` check it for `node.expression` only when the `node` is a `NewExpression`.

Probably orders of magnitude less expensive than the tree exploration done in the step functions code or the re-computation of lexical scope.